src/core/nvector.cc

```diff
@@ -0,0 +1,451 @@
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program; if not, write to the Free Software
+    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+ */
+
+// https://en.wikipedia.org/wiki/N-vector
+// http://www.navlab.net/Publications/A_Nonsingular_Horizontal_Position_Representation.pdf
+
+#include <cfloat>
+#include <cmath>
+#include <cstdio>
+#include <iostream>
+#include <utility>
+
+#include "defs.h"
+#include "nvector.h"
+#include "vector3d.h"
+
+namespace gpsbabel
+{
+
+LatLon::LatLon(double latitude, double longitude)
+{
+  lat = latitude;
+  lon = longitude;
+}
+
+NVector::NVector(double latitude_degrees, double longitude_degrees)
+{
+  // this implements equation 3.
+
+  // The coordinate system choice matches
+  // A_Nonsingular_Horizontal_Position_Representation.pdf
+  // The E frame:
+  // from the earths center
+  // x points to North pole
+  // y points towards latitude 0, longitude +90 (east)
+  // z points to latitude 0, longitude +180, the antimeridian
+  double latitude_radians = latitude_degrees * M_PI/180.0;
+  double longitude_radians = longitude_degrees * M_PI/180.0;
+  x = sin(latitude_radians);
+  y = sin(longitude_radians)*cos(latitude_radians);
+  z = -cos(longitude_radians)*cos(latitude_radians);
+}
+
+NVector::NVector(const Vector3D& v)
+{
+  x = v.getx();
+  y = v.gety();
+  z = v.getz();
+}
+
+std::pair<NVector, double> PVector::toNVectorAndHeight() const
+{
+  // this implements equation 23.
+  constexpr double a = WGS84_SEMI_MAJOR_AXIS_METERS;
+  constexpr double a2 = a * a;
+  constexpr double e2 = WGS84_ECCENTRICITY_SQUARED;
+  constexpr double e4 = e2 * e2;
+
+  double px2 = getx() * getx();
+  double py2 = gety() * gety();
+  double pz2 = getz() * getz();
+
+  double q = ((1.0-e2)/(a2)) * px2;
+  double p = (py2 + pz2) / (a2);
+  double r = (p+q-e4) / 6.0;
+  double s = (e4*p*q) / (4.0*r*r*r);
+  double t = cbrt(1.0 + s + sqrt(s*(2.0+s)));
+  double u = r * (1.0 + t + (1.0/t));
+  double v = sqrt(u*u + e4*q);
+  double w = e2 * ((u+v-q)/(2.0*v));
+  double k = sqrt(u+v+w*w) - w;
+  double d = k*sqrt(py2 + pz2) / (k+e2);
+  double sf = 1.0 / (sqrt(d*d + px2));
+  double sf2 = k / (k+e2);
+
+  double height = ((k+e2-1.0)/k) * sqrt(d*d+px2);
+  double nx = sf * getx();
+  double ny = sf * sf2 * gety();
+  double nz = sf * sf2 * getz();
+
+  return {Vector3D(nx, ny, nz), height};
+}
+
+double NVector::latitude() const
+{
+  // this implements equation 6.
+  double latitude_radians = atan2(x, sqrt(y*y + z*z));
+  double latitude_degrees = latitude_radians * 180.0/M_PI;
+  return latitude_degrees;
+}
+
+double NVector::longitude() const
+{
+  // this implements equation 5.
+  double longitude_radians = atan2(y, -z);
+  double longitude_degrees = longitude_radians * 180.0/M_PI;
+  return longitude_degrees;
+}
+
+// great circle distance in radians
```

Comment:
Sentence casing and punctuation, please.

Reply:
I pushed another commit with improvements to capitalization and spelling.